### ISSUE TITLE
PostService GetByIdAsync does not always return post

### ISSUE THREAD

**Issue by BryanEuton** [PiranhaCMS/piranha.core#1867]
If a post is in the cache and it is assignable from DynamicPost, then the Post will not return from PostService.GetByIdAsync<T>.

The problem is in this method:
![image](https://user-images.githubusercontent.com/13066934/166503140-d57d7051-782d-49a7-9d62-84c817d6c7fc.png)

On the line `model = _cache?.Get<PostBase>(id.ToString());`, the model is received from the cache but it is of type PostBase not T.  So null is returned when the model is checked if it is of type "T".  

I found this because I created a custom class like PostInfo, but with my additional fields.   

To correct this issue, `model = _cache?.Get<PostBase>(id.ToString());` should be changed to `model = _cache?.Get<T>(id.ToString());`.  I don't believe that there would be any issues from doing this.

**Comment by tidyui** [comment]
Hi there! Let me clarify how the code works. Posts are only stored, and retrieved from cache **IF**:

1. `T` is exactly `Piranha.Models.PostInfo`
2. Or if `T` is a type derived from `Piranha.Models.PostBase` but **NOT** from `Piranha.Models.DynamicPost`

This means that dynamic models are never cached as they are only used internally from within the manager UI. The if statement you're talking about (`model is T`) will not fail because posts are fetched from the cache as `PostBase` as `T` will always derive from `PostBase`. 

As I don't have much information about your source code the next part will be some qualified guessing :) The only way this statement should be able to fail is if you request **the same page instance** (i.e Guid id) as different CLR types from different parts of your code, for example having the following structure.

~~~ csharp
public class MyPostPreview : Post<MyPostPreview>
{
   ...
}

[PostType(...)]
public class MyPost : MyPostPreview
{
   ...
}

var preview = await api.Posts.GetByIdAsync<MyPostPreview>(id);
var post = await api.Post.GetByIdAsync<MyPost>(id);
~~~

The second call will fail as the first call will add a model of the type `MyPostPreview` into the cache with the key `id`, the second time it is loaded as `MyPost` but the object returned from cache will not match the requested model type. 

Simply put, the services was never designed to support this sort of scenario, and even if **one** problem can be overcome by changing the code as you suggest, the problem then arises that only one of the model types would be cached and the other would always be retrieved from the database resulting on performance drops. Also, it is not deterministic as the type that will work will be 100% dependend on the first type that is requested.

If I didn't understand your problem correctly, please provide more information about your use-case.

**Comment by tidyui** [comment]
Also, as the entire transformed model (in my example `MyPost`) will be stored in cache there is no performance hit on getting the full model as it will be ready in cache. If you would like methods returning view models this should be added to another service, for example.

~~~ csharp
public class ViewModelService
{
   private readonly IApi _api;
   private readonly ICache _cache;

   public async Task<T> GetPostAsync<T>(Guid id) where T : PostBase
   {
      var model = _cache.Get<T>($"{ typeof(T).Name }_{ id }");

      if (model == null)
      {
         var fullModel = await _api.Posts.GetById<PostBase>(id);

         model = TransformFullModelToViewModel<T>(fullModel);

         _cache.Set($"{ typeof(T).Name }_{ id }", model);
      }
      return model;
   }
}
~~~

**NOTE:** I wrote this code in this texteditor as an example and has not compiled it in any way as it is not complete. Also, this kind of solution would need to listen to the available hooks for posts so that it can clear objects from its cache on `Save` and `Delete`.

Regards

**Comment by BryanEuton** [comment]
Sorry for the delay.  

I followed the StandardPost example:
`
    [PostType(Title = "Standard event")]
    [ContentTypeRoute(Title = "Default", Route = "/...")]
    public class StandardPostEvent : Post<StandardPostEvent>
    {
...
    }
`
the post api retrieves all of the posts on the first try.  Once the post is in cache though, that post won't be rendered in the list anymore though.  The problem is the _cache?.Get<PostBase>... line.  When it pulls the item from cache, it is of type <PostBase> since that is the brackets.  Later in this method, you return the item if it is of Type "T".   
In my opinion, we either need to modify the _cache?.Get<PostBase> to be _cache?.Get<T> or after getting it from cache see if the model is of Type "T" so that we can pull it from the database when it isn't.  
My preference is the first way obviously since it is being cached (aka model isn't null).  Here is the workaround that I had to use due to this bug.

`
        public async Task<PostArchive<T>> GetByIdAsync<T>(Guid archiveId, int? currentPage = 1,
            Guid? categoryId = null, Guid? tagId = null, int? year = null, int? month = null, int? pageSize = null)
            where T : StandardPostEvent
        {
            var model = new PostArchive<T>();
....
            // Get the id of the current posts
            var posts = await _repo.GetPosts(archiveId, pageSize.Value, model.CurrentPage, categoryId, tagId, year, month).ConfigureAwait(false);

            // Get the posts
            foreach (var postId in posts)
            {
                var post = await _postService.GetByIdAsync<T>(postId);

                if (post != null)
                {
                    model.Posts.Add(post);
                }
                else
                {
//This happens when the model is in cache.  Retrieve it straight from cache.
                    var cachedPost = await GetPostByIdAsync<T>(postId, new List<PageInfo>());
                    if(cachedPost != null)
                    {
                        model.Posts.Add(cachedPost);
                    }
                }
            }
            return model;
        }`
Here is the GetPostByIdAsync<T> called from above.  (Looks super similar to the Piranha code... :)  Only difference is the _cache?.Get<T> line.
`public async Task<T> GetPostByIdAsync<T>(Guid id, IList<PageInfo> blogPages) where T : PostBase
        {

            PostBase model = null;

            if (typeof(T) == typeof(PostInfo))
            {
                model = _cache?.Get<PostInfo>($"PostInfo_{id.ToString()}");
            }
            else if (!typeof(DynamicPost).IsAssignableFrom(typeof(T)))
            {
                model = _cache?.Get<T>(id.ToString());

                if (model != null)
                {
                    await _factory.InitAsync(model, App.PostTypes.GetById(model.TypeId));
                }
            }


            if (model == null)
            {
                model = await _postRepository.GetById<T>(id).ConfigureAwait(false);

                if (model != null)
                {
                    var blog = blogPages.FirstOrDefault(p => p.Id == model.BlogId);

                    if (blog == null)
                    {
                        blog = await _pageService.GetByIdAsync<PageInfo>(model.BlogId).ConfigureAwait(false);
                        blogPages.Add(blog);
                    }

                    await OnLoadAsync(model, blog).ConfigureAwait(false);
                }
            }

            if (model != null && model is T)
            {
                return (T)model;
            }
            return null;
        }
`

**Comment by tidyui** [comment]
Out of curiosity, as this code has always been the same, what cache are you using?

**Comment by tidyui** [comment]
@BryanEuton This is the whole point I’m trying to make. The model returned will never be of the type “PostBase” as this is an abstract base class that can’t be an instance. If the system works as expected the model will always be of the type T which is the model type derived from PostBase.

The exact same type of code is executing for the page service, see https://github.com/PiranhaCMS/piranha.core/blob/master/core/Piranha/Services/PageService.cs#L263. If you’re not experiencing the same issues with the page cache the error is within the application, I just can’t figure out what it is.

Have you inspected the model that is returned from the cache and how it matches the expected model?

Regards

**Comment by BryanEuton** [comment]
@tidyui  Sorry I can't seem to reproduce this anymore.  I'll close this issue.
